### PR TITLE
qgnomeplatform: hardcode gsettings schemas

### LINKED ISSUE #81866: qgnomeplatform requires gsettings-desktop-schemas
Following the last couple, I can now reproduce the issue again. I'm running NixOS off of [this branch](https://github.com/lovesegfault/nixpkgs/commits/nixpkgs-lovesegfault), which is just `nixos-unstable` with the previous too patches, that were still in staging`, applied.

With that branch I get the following when I try to launch zoom:
```
❯ zoom-us
ZoomLauncher started.
cmd line:
CreateReportChannel bp_server_fd=4
$HOME = /home/bemeurer
export SSB_HOME=/home/bemeurer/.zoom; export QSG_INFO=1; export LD_LIBRARY_PATH=/nix/store/ysp9ql4l038m13gd3wyn5gr66xg60lc7-zoom-us-3.5.361976.0301/share/zoom-us; export BREAKPAD_CLIENT_FD=3; /nix/store/ysp9ql4l038m13gd3wyn5gr66xg60lc7-zoom-us-3.5.361976.0301/share/zoom-us/zoom ""
zoom started.
Client: Breakpad is using Client-Server Mode! client fd = 3

(process:24790): GLib-GIO-ERROR **: 20:13:11.433: Settings schema 'org.gnome.desktop.wm.preferences' is not installed
ZoomLauncher: new dump generated from pid 24790, path: /home/bemeurer/.zoom/logs/39619221-5fab-460d-0e59b3a3-b25c461f.dmp
$HOME = /home/bemeurer
sh: line 1: 24790 Trace/breakpoint trap   (core dumped) /nix/store/ysp9ql4l038m13gd3wyn5gr66xg60lc7-zoom-us-3.5.361976.0301/share/zoom-us/zoom ""
success to create child process,status is 34048.
zoom exited normally.
Something went wrong while running zoom,exit code is 133.
ZoomLauncher exit.
```

Edit: See https://github.com/NixOS/nixpkgs/issues/81866#issuecomment-602851994 https://github.com/NixOS/nixpkgs/issues/81866#issuecomment-599676558

### PR DESCRIPTION
**Pull request by worldofpeace**
Fixes #81866

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by lovesegfault — APPROVED** [review]
Works flawlessly!

File: pkgs/development/libraries/qgnomeplatform/hardcode-gsettings.patch
Comment (jtojnar) [inline pkgs/development/libraries/qgnomeplatform/hardcode-gsettings.patch]: The issue with this is that it does not free the schema source and schemas. That is too advanced for initializer lists so we need to move all of these into the body of the constructor (and then use [`g_autoptr`](https://developer.gnome.org/glib/stable/glib-Miscellaneous-Macros.html#g-autoptr) on those). Fortunately, we are only dealing with object’s own properties so it should be fine.

Comment (jtojnar) [reply pkgs/development/libraries/qgnomeplatform/hardcode-gsettings.patch]: Also, it looks like the original code is not correctly disposing of [`GSettings`](https://developer.gnome.org/gio/stable/GSettings.html) instances either. They should call [`g_object_unref`](https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html#g-object-unref) in the [destructor](https://github.com/FedoraQt/QGnomePlatform/blob/a4a5788e754c0d55980454a60a04963a2ca31aa6/common/gnomehintssettings.cpp#L148).

**Review by jtojnar — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/qgnomeplatform/hardcode-gsettings.patch
Comment (jtojnar) [reply pkgs/development/libraries/qgnomeplatform/hardcode-gsettings.patch]: Opened https://github.com/FedoraQt/QGnomePlatform/pull/57

**Comment by jtojnar** [comment]
Moved the schema initialization from initializer lists inside the constructor, where the auxiliary objects will be unref’d upon leaving the stack frame (thanks to the magic of `g_autoptr`). Not sure if fetching also https://github.com/FedoraQt/QGnomePlatform/commit/db2afbed7cdcc63b6d27341bd6920a84e5e00598 is worth it.

**Comment by worldofpeace** [comment]
> Moved the schema initialization from initializer lists inside the constructor, where the auxiliary objects will be unref’d upon leaving the stack frame (thanks to the magic of `g_autoptr`). Not sure if fetching also [FedoraQt/QGnomePlatform@db2afbe](https://github.com/FedoraQt/QGnomePlatform/commit/db2afbed7cdcc63b6d27341bd6920a84e5e00598) is worth it.

Thanks soo much :sparkles:

**Comment by worldofpeace** [comment]
eval problem is unrelated https://github.com/NixOS/nixpkgs/issues/83929

**Comment by worldofpeace** [comment]
backported to 20.03 in 33c19df80d98ee48c8bb23e0736d8cc9be04ab31

**Comment by jtojnar** [comment]
Note that I did not test it other than by building it.

**Comment by worldofpeace** [comment]
> Note that I did not test it other than by building it.

Ohh, I should have mentioned it before I merged. I did test it on a Pantheon system, but I actually couldn't reproduce the issue before and after when running a program in a shell with unset XDG_DATA_DIRS. I guess I'll leave that up to @lovesegfault to find out. (I tested it with vlc)

**Comment by gnidorah** [comment]
@worldofpeace It fixes issue for me

**Comment by lovesegfault** [comment]
It works for myself as well :)

**Comment by worldofpeace** [comment]
Ohh, forgot to backport. Will do this.

**Comment by gnidorah** [comment]
Found an issue with this solution:
1) Start i3wm
2) Launch VLC via dmenu
3) Click Media - Open file (Ctrl + O)
Boom - crash!

**Comment by worldofpeace** [comment]
That is actually within gtk. So it isn't related to this exact fix.

@jtojnar Could we just hardcode those one's in gtk?

**Comment by jtojnar** [comment]
Yes, that's what I wanted to do, just did not get to it yet. Though, I believe GTK also depends on gsettings-desktop-schemas.

**Pull request merged**
